pilot/pkg/serviceregistry/kube/controller/filter/discoverynamespaces.go

```diff
@@ -0,0 +1,54 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filter
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	v1 "k8s.io/client-go/listers/core/v1"
+
+	"istio.io/pkg/log"
+)
+
+const (
+	// PilotDiscoveryLabelName is the name of the label used to indicate a namespace for discovery
+	PilotDiscoveryLabelName = "istio-discovery"
```

Comment:
While the current label is istio-injection, its mostly that out of legacy. Any new labels should be `*.istio.io/*` format to conform to k8s best practice

Reply:
Any opinions on the exact label string? Would `discovery.istio.io/enable` be conventional?

Reply:
There was [some discussion](https://github.com/istio/istio/issues/26679#issuecomment-742642458) about allowing more than one istiod to managed (namespace-filtered) meshes in the same cluster. This implementation doesn't seem to support that. If it is required, the value of this label needs to be a meshId, not just boolean.

Reply:
If I understand correctly, the recently proposed `discoverySelectors` option should enable multiple istiod's to manage mesh in the same cluster with non overlapping selectors. 
@harveyxia: [NewNamespaceController](https://github.com/istio/istio/blob/master/pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go#L58) should also honor this selector configuration and insert `istio-ca-root-cert` config map to only filtered namespaces.

Reply:
NamespaceController should also use DiscoveryNamespacesFilter
@harveyxia do you plan to update NamespaceController?

Reply:
@l8huang The original motivation for this implementation was primarily for performance improvement (filtering out irrelevant services, pods, and endpoints) , so we kept the `DiscoveryNamespacesFilter` scoped to the main service registry controller. With some refactoring (making `DiscoveryNamespacesFilter` self contained and exposing methods for registering handlers for meshConfig and namespace events) it can be extended for use with other components.

However, before we do any implementation, we should determine whether this feature is the right abstraction for enabling multiple istiod deployments (which is an orthogonal concern to performance), or if there are other designs / existing components that would be better suited for this use case. @howardjohn @costinm any thoughts?